spring-web/src/main/java/org/springframework/http/client/reactive/HttpComponentsClientHttpRequest.java

```diff
@@ -0,0 +1,152 @@
+
+import static org.springframework.http.MediaType.ALL_VALUE;
+
+/**
+ * {@link ClientHttpRequest} implementation for the Apache HttpComponents HttpClient 5.x.
+ *
+ * @author Martin Tarjányi
+ * @since 5.3
+ * @see <a href="https://hc.apache.org/index.html">Apache HttpComponents</a>
+ */
+class HttpComponentsClientHttpRequest extends AbstractClientHttpRequest {
+
+	private final HttpRequest httpRequest;
+
+	private final DataBufferFactory dataBufferFactory;
+
+	private final HttpClientContext context;
+
+	private Flux<ByteBuffer> byteBufferFlux;
+
+	private long contentLength = -1;
+
+
+	public HttpComponentsClientHttpRequest(HttpMethod method, URI uri, HttpClientContext context,
+			DataBufferFactory dataBufferFactory) {
+
+		this.context = context;
+		this.httpRequest = new BasicHttpRequest(method.name(), uri);
+		this.dataBufferFactory = dataBufferFactory;
+	}
+
+
+	@Override
+	public HttpMethod getMethod() {
+		return HttpMethod.resolve(this.httpRequest.getMethod());
+	}
+
+	@Override
+	public URI getURI() {
+		try {
+			return this.httpRequest.getUri();
+		}
+		catch (URISyntaxException ex) {
+			throw new IllegalArgumentException("Invalid URI syntax.", ex);
+		}
+	}
+
+	@Override
+	public DataBufferFactory bufferFactory() {
+		return this.dataBufferFactory;
+	}
+
+	@Override
+	public Mono<Void> writeWith(Publisher<? extends DataBuffer> body) {
+		return doCommit(() -> {
+			this.byteBufferFlux = Flux.from(body).map(DataBuffer::asByteBuffer);
+			return Mono.empty();
+		});
+	}
+
+	@Override
+	public Mono<Void> writeAndFlushWith(Publisher<? extends Publisher<? extends DataBuffer>> body) {
+		return writeWith(Flux.from(body).flatMap(p -> p));
+	}
+
+	@Override
+	public Mono<Void> setComplete() {
+		return doCommit();
+	}
+
+	@Override
+	protected void applyHeaders() {
+		HttpHeaders headers = getHeaders();
+		this.contentLength = headers.getContentLength();
+
+		headers.entrySet()
+				.stream()
+				.filter(entry -> !HttpHeaders.CONTENT_LENGTH.equals(entry.getKey()))
+				.forEach(entry -> entry.getValue().forEach(v -> this.httpRequest.addHeader(entry.getKey(), v)));
+
+		if (!this.httpRequest.containsHeader(HttpHeaders.ACCEPT)) {
+			this.httpRequest.addHeader(HttpHeaders.ACCEPT, ALL_VALUE);
+		}
+	}
+
+	@Override
+	protected void applyCookies() {
+		if (getCookies().isEmpty()) {
+			return;
+		}
+
+		CookieStore cookieStore = this.context.getCookieStore();
+
+		getCookies().values()
+				.stream()
+				.flatMap(Collection::stream)
+				.forEach(cookie -> {
+					BasicClientCookie clientCookie = new BasicClientCookie(cookie.getName(), cookie.getValue());
+					clientCookie.setDomain(getURI().getHost());
+					clientCookie.setPath(getURI().getPath());
+					cookieStore.addCookie(clientCookie);
+				});
+	}
+
+	public HttpRequest getHttpRequest() {
+		return this.httpRequest;
+	}
+
+	public Flux<ByteBuffer> getByteBufferFlux() {
+		return this.byteBufferFlux;
+	}
+
+	public long getContentLength() {
+		return this.contentLength;
+	}
```

Comment:
These getters should be removed, in favor of having a `toRequestProducer` method. See above.